### PR TITLE
Early load rocksdb plugin if ddse is rocksdb

### PR DESCRIPTION
**Pull request by sunshine-Chun**
Summary: 
Load rocksdb plugin in early stage if ddse is rocksdb

Test Plan: MTR

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test [hunk 1]
```diff
@@ -0,0 +1,34 @@
+--source include/have_rocksdb.inc
+
+# Set the default ddse to Rocksdb
+# Stimulate bootstrap by starting mysqld with --initialize
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test]:
```suggestion
# Simulate bootstrap by starting mysqld with --initialize
```

But this is not a "simulation" - this is an actual bootstrap on some instance

File: mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+--remove_file $MYSQLD_LOG
+--force-rmdir $DDIR
+
+--echo #
+--echo # Cleanup
+--echo #
+--echo # Restarting the server
+--source include/start_mysqld.inc
+
+SELECT count(*) FROM INFORMATION_SCHEMA.PLUGINS where PLUGIN_NAME like '%ROCKSDB%';
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test]: Redundant?

File: sql/sql_plugin.cc [hunk 3]
```diff
@@ -1577,20 +1577,23 @@ bool plugin_register_builtin_and_init_core_se(int *argc, char **argv) {
       if (register_builtin(plugin, &tmp, &plugin_ptr)) goto err_unlock;
 
       /*
-        Only initialize daemon_keyring_proxy, MyISAM, InnoDB and CSV at this
-        stage. Note that when the --help option is supplied,
-        daemon_keyring_proxy and InnoDB are not initialized because the plugin
-        table will not be read anyway, as indicated by the flag set when the
-        plugin_init() function is called.
+        Only initialize daemon_keyring_proxy, MyISAM, InnoDB, CSV and ROCKSDB
+        (only when defaul dd storage engine is Rocksdb) at this stage. Note that
```
Comment (laurynas-biveinis) [inline sql/sql_plugin.cc]: Inconsistent "ROCKSDB/Rocksdb" case

File: sql/sql_plugin.cc [hunk 4]
```diff
@@ -1577,20 +1577,23 @@ bool plugin_register_builtin_and_init_core_se(int *argc, char **argv) {
...
           &my_charset_latin1, plugin->name, "daemon_keyring_proxy_plugin");
       bool is_myisam =
           !my_strcasecmp(&my_charset_latin1, plugin->name, "MyISAM");
       bool is_innodb =
           !my_strcasecmp(&my_charset_latin1, plugin->name, "InnoDB");
+      bool is_rocksdb =
+          !my_strcasecmp(&my_charset_latin1, plugin->name, "ROCKSDB");
       if ((!is_daemon_keyring_proxy || is_help_or_validate_option()) &&
           !is_myisam && (!is_innodb || is_help_or_validate_option()) &&
+          (!is_rocksdb || default_dd_storage_engine != DEFAULT_DD_ROCKSDB) &&
```
Comment (laurynas-biveinis) [inline sql/sql_plugin.cc]: Not sure what should we do about this at this stage - we will need to implement cross-grade between different DD engines. Might be OK for now though

File: sql/sql_plugin.cc [hunk 5]
```diff
@@ -3237,8 +3240,9 @@ static int construct_options(MEM_ROOT *mem_root, st_plugin_int *tmp,
       (char *)mem_root->Alloc(plugin_name_len + plugin_dash.length + 1);
   strxmov(plugin_name_with_prefix_ptr, plugin_dash.str, plugin_name_ptr, NullS);
 
-  if (tmp->load_option != PLUGIN_FORCE &&
-      tmp->load_option != PLUGIN_FORCE_PLUS_PERMANENT) {
+  if ((tmp->load_option != PLUGIN_FORCE &&
+       tmp->load_option != PLUGIN_FORCE_PLUS_PERMANENT) ||
+      (strcmp(plugin_name, "ROCKSDB") == 0)) {
```
Comment (laurynas-biveinis) [inline sql/sql_plugin.cc]: Why is this needed?

File: storage/rocksdb/rdb_datadic.cc [hunk 6]
```diff
@@ -4342,6 +4342,12 @@ int Rdb_validate_tbls::add_table(Rdb_tbl_def *tdef) {
 */
 bool Rdb_validate_tbls::validate(void) {
   THD *const thd = my_core::thd_get_current_thd();
+
+  if(!thd) {
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: Formatting (`clang-format`)

File: storage/rocksdb/CMakeLists.txt [hunk 7]
```diff
@@ -266,7 +266,7 @@ IF (ROCKSDB_DYNAMIC_PLUGIN)
   MYSQL_ADD_PLUGIN(rocksdb ${ROCKSDB_SOURCES} STORAGE_ENGINE DEFAULT
     MODULE_ONLY LINK_LIBRARIES ${rocksdb_static_libs})
 ELSE()
-  MYSQL_ADD_PLUGIN(rocksdb_se ${ROCKSDB_SOURCES} STORAGE_ENGINE DEFAULT
+  MYSQL_ADD_PLUGIN(rocksdb_se ${ROCKSDB_SOURCES} STORAGE_ENGINE MANDATORY
```
Comment (laurynas-biveinis) [inline storage/rocksdb/CMakeLists.txt]: The ability to build RocksDB as a dynamic plugin remains - which is incompatible with native DD and is likely to be broken by the `sql_plugin.cc` changes. As a minimum, please test this PR with `-DWITH_ROCKSDB_DYNAMIC_PLUGIN=ON` (not sure if that's the exact CMake option name), and then please either make any changes to make it work, either remove the dynamic plugin option. But the latter may break some user scripts for a feature that does not exist yet.

File: sql/sql_plugin.cc (hunk 4 above)
Comment (luqun) [reply sql/sql_plugin.cc]: yes. I think this PR only enable load rocksdb plugin before DD initialize..

File: mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+--source include/shutdown_mysqld.inc
+--echo # Server is down
+
+--echo #
+--echo # Try --initialize
+--echo #
+
+--echo # Run the server with --initialize
+--exec $MYSQLD $extra_args --initialize --datadir=$DDIR --log-error-verbosity=1 > $MYSQLD_LOG 2>&1
+
```
Comment (luqun) [inline mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test]: could you check "RocksDB:Init column families" or similar string in MYSQLD_LOG?

File: storage/rocksdb/CMakeLists.txt (hunk 7 above)
Comment (sunshine-Chun) [reply storage/rocksdb/CMakeLists.txt]: In the diff tool, we have test set for both dynamicRocksdb and static rocksdb. From the test result, most of them passed. There are few test case failure. But I don't think it's related to the change. @luqun Could you please help to verify it's not related to this diff's change?

File: sql/sql_plugin.cc (hunk 5 above)
Comment (sunshine-Chun) [reply sql/sql_plugin.cc]: This is needed when I run setup command in dev server to build a database. Without this change, the setup script will fail.

Comment (laurynas-biveinis) [reply sql/sql_plugin.cc]: What is the command that the setup script executes and what error is it getting?

Comment (sunshine-Chun) [reply sql/sql_plugin.cc]: It's about handling the command options. When executing mysqld command in the setup script, we pass --rocksdb as a command parameter. After changing Rocksdb as mandatory, this option is considered loose anymore. And will throw an error about unknown option and fail the mysqld. The change here helps to add the option to option list to pass the option check.

File: mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+--echo #
+--echo # Try --initialize
+--echo #
+
+--echo # Run the server with --initialize
+--exec $MYSQLD $extra_args --initialize --default_dd_storage_engine=RocksDB --datadir=$DDIR --log-error-verbosity=3 > $MYSQLD_LOG 2>&1
+
+--let OUTF = $MYSQLD_LOG
+
+perl;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/early_load_rocksdb_plugin.test]: Please look into `assert_grep.inc`, specifically its `assert_only_after` parameter

File: sql/sql_plugin.cc (hunk 5 above)
Comment (laurynas-biveinis) [reply sql/sql_plugin.cc]: I don't like the special casing here, but what has to be done has to be done. Perhaps the scripts should no longer pass `--rocksdb` (as they are not passing `--innodb` currently, right?), and then later this could be removed.

File: sql/sql_plugin.cc [hunk 10]
```diff
@@ -3237,8 +3240,12 @@ static int construct_options(MEM_ROOT *mem_root, st_plugin_int *tmp,
       (char *)mem_root->Alloc(plugin_name_len + plugin_dash.length + 1);
   strxmov(plugin_name_with_prefix_ptr, plugin_dash.str, plugin_name_ptr, NullS);
 
-  if (tmp->load_option != PLUGIN_FORCE &&
-      tmp->load_option != PLUGIN_FORCE_PLUS_PERMANENT) {
+  /* After changing static ROCKSDB plugin as mandatory, the --rocksdb option is
+   * not considered as loose. Add the option in the option list to avoid unknown
```
Comment (laurynas-biveinis) [inline sql/sql_plugin.cc]: Not a common MySQL comment style, the common ones are `//` and `/* */` without leading asterix in the intermediate lines

File: storage/rocksdb/rdb_datadic.cc [hunk 11]
```diff
@@ -4342,6 +4342,15 @@ int Rdb_validate_tbls::add_table(Rdb_tbl_def *tdef) {
 */
 bool Rdb_validate_tbls::validate(void) {
   THD *const thd = my_core::thd_get_current_thd();
+
+  /* When ddse is ROCKSDB, we are not able to validate the Rocksdb tables since
+   * we load rocksdb plugin before initializing dd. So skip validation here. */
```
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.cc]: Comment style

File: sql/sql_plugin.cc (hunk 5 above)
Comment (luqun) [reply sql/sql_plugin.cc]: @sunshine-Chun , if we revert this change, and if MTR passes, we can upgrade script to remove that --rocksdb instead.

Comment (sunshine-Chun) [reply sql/sql_plugin.cc]: Sounds good. Let me remove this line and see if MTR test passes

File: storage/rocksdb/CMakeLists.txt (hunk 7 above)
Comment (luqun) [reply storage/rocksdb/CMakeLists.txt]: I will take a look this part to see whether it works with rocksdb dynamic plugin..

File: sql/sql_plugin.cc (hunk 5 above)
Comment (sunshine-Chun) [reply sql/sql_plugin.cc]: Looks like the MTR passed. Will remove this line.
